R/eml_to_spice.R

```diff
@@ -0,0 +1,294 @@
+get_entities <- function(eml,
+                         entities = c("dataTable", "spatialRaster", "spatialVector", "storedProcedure", "view", "otherEntity"),
+                         level_id = "entityName"){
+    entities <- entities[entities %in% names(eml$dataset)]
+
+    #look for specific fields to determine if the entity needs to be listed ("boxed") or not
+    level_cond <- paste0("~", paste(sprintf("!is.null(.x$%s)", level_id), collapse = " | "))
+    purrr::map(entities, ~eml2::eml_get(eml, .x)) %>%
+        # restructure so that all entities are at the same level
+        # use level id to determine if .x should be listed or not
+        purrr::map_if(eval(parse(text = level_cond)), list) %>%
+        unlist(recursive = FALSE)
+}
+
+get_access_spice <- function(x){
+    x %>%
+        unlist() %>%
+        tibble::enframe() %>%
+        dplyr::mutate(name = dplyr::case_when(
+            grepl("objectName", name) ~ "fileName",
+            grepl("entityName", name) ~ "name",
+            grepl("url", name) ~ "contentUrl",
+            grepl("formatName", name) ~ "fileFormat"
+        )) %>%
+        stats::na.omit() %>%
+        filter(value != "download") %>% #often also included as url
+        tidyr::spread(name, value)
+}
+
+#' Get access from EML
+#'
+#' Return EML access in the dataspice access.csv format.
+#'
+#' @param eml (emld) an EML object
+#' @param path (character) folder path for saving the table to disk
+#'
+#' @export
+#' @import eml2
+#'
+#' @examples
+#' \dontrun{
+#' eml_path <- system.file("example-dataset/broodTable_metadata.xml", package = "dataspice")
+#' eml <- read_eml(eml_path)
+#' es_access(eml)
+#' }
+
+es_access <- function(eml, path = NULL) {
+    entities <- get_entities(eml)
+    access_entities <- lapply(entities, get_access_spice)
+
+    out <- dplyr::bind_rows(access_entities)
+
+    #reorder
+    fields <- c("fileName", "name", "contentUrl", "fileFormat")
+    out <- out[, fields[fields %in% colnames(out)]]
+
+    if(!is.null(path)){
+        if(!dir.exists(path)){
+            dir.create(path)
+        }
+        readr::write_csv(out, file.path(path, "access.csv"))
+    }
+
+    return(out)
+}
+
+get_attributes_spice <- function(x) {
+  #reformat attributes to tabular format specified in dataspice
+  #input a dataTable or otherEntity
+
+  objName <- eml2::eml_get(x, "objectName")
+  objName <- ifelse(length(objName) == 2, objName[[1]], NA)
+
+  attrList <- eml2::eml_get(x, "attributeList")
+
+  if(length(attrList) <= 1){
```

Comment:
I'm curious about this branch condition. Why not `length(attrList) == 0`

Reply:
I had to use `<= 1` due to a recent change to `eml2`/`emld`, which adds context as an extra list item.  I didn't totally understand what was going at the time, so it's admittedly not the neatest solution.  Probably better is to remove any `@context` list items and then do `length(attrList) == 0`.  @cboettig do you have a better solution for this?

@amoeba If you want to take a look, make sure you have the latest versions of eml2/emld installed, and try this:
```
library(eml2)
f <- system.file("extdata", "example.xml", package = "emld")
eml <- read_eml(f)
sn <- eml_get(eml, "surName")
```

Reply:
Ah, makes sense. There might be a way to de-JSONLD the list or something but this is fine.